### PR TITLE
Add ability to bake BHoMObjects and BHoMGeometry

### LINKED ISSUE #549: Add ability to Bake BHoMGeometry as well as BHoMObejcts
#### Description:

Adding Feature to be able to Bake BHoMGeometry.

Also, adding feature to Bake any BHoMObject that has a `Geometry()` method implemented (any Previewable object). Can then assign the name of the object (if any) to the baked obejct

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

Closes #549

Did some quick digging after the workshop, and realised this was relatively simple, and actually already half-solved for geometry, so went ahead and just fixed it. Hope this does not derail anything @adecler , if so, please feel free to close/park.

- Making it possible to bake out BHoMGeometry
   - One half of this was actually already working. You actually were able to bake BHoMGeometry from a Data parameter. For this, Just needed to make the BHoMGeometry parameter bake aware as well.
- Adding Baking capability to the BHoMObject Goo and paramter
- Adding Baking capability to the IObjectGoo and paramter

### Test files

Try baking some random BHoMObejcts and geometries. Should bake any BHoMObejct that is previewable, no others.

### CODE REVIEW

File: Grasshopper_UI/Components/Parameters/Param_BHoMGeometry.cs [hunk 1]
```diff
@@ -38,15 +42,42 @@ public class Param_BHoMGeometry : BHoMParam<Engine.Grasshopper.Objects.GH_IBHoMG
...
+        }
+
+        /*******************************************/
+
+        public void BakeGeometry(RhinoDoc doc, ObjectAttributes att, List<Guid> obj_ids)
+        {
+            foreach (Engine.Grasshopper.Objects.GH_IBHoMGeometry item in this.VolatileData.AllData(true))
+            {
+                Guid guid;
+                if (item.BakeGeometry(doc, att, out guid))
```
Comment (adecler) [inline Grasshopper_UI/Components/Parameters/Param_BHoMGeometry.cs:68]: Probably a good idea to check that `item` is not null

File: Grasshopper_UI/Components/Parameters/Param_BHoMGeometry.cs [hunk 2]
```diff
@@ -38,15 +42,42 @@ public class Param_BHoMGeometry : BHoMParam<Engine.Grasshopper.Objects.GH_IBHoMG
...
+        /***************************************************/
+        /**** IGH_BakeAwareObject methods               ****/
+        /***************************************************/
+
+        public void BakeGeometry(RhinoDoc doc, List<Guid> obj_ids)
+        {
+            foreach (Engine.Grasshopper.Objects.GH_IBHoMGeometry item in this.VolatileData.AllData(true))
+            {
+                Guid guid;
+                if (item.BakeGeometry(doc, null, out guid))
```
Comment (adecler) [inline Grasshopper_UI/Components/Parameters/Param_BHoMGeometry.cs:56]: Probably a good idea to check that `item` is not null

File: Grasshopper_UI/Components/Parameters/Param_BHoMObject.cs [hunk 3]
```diff
@@ -38,6 +42,33 @@ public class Param_BHoMObject : BHoMParam<Engine.Grasshopper.Objects.GH_BHoMObje
...
+        /***************************************************/
+        /**** IGH_BakeAwareObject methods               ****/
+        /***************************************************/
+
+        public void BakeGeometry(RhinoDoc doc, List<Guid> obj_ids)
+        {
+            foreach (Engine.Grasshopper.Objects.GH_BHoMObject item in this.VolatileData.AllData(true))
+            {
+                Guid guid;
+                if (item.BakeGeometry(doc, null, out guid))
```
Comment (adecler) [inline Grasshopper_UI/Components/Parameters/Param_BHoMObject.cs:56]: Probably a good idea to check that `item` is not null

File: Grasshopper_UI/Components/Parameters/Param_BHoMObject.cs [hunk 4]
```diff
@@ -38,6 +42,33 @@ public class Param_BHoMObject : BHoMParam<Engine.Grasshopper.Objects.GH_BHoMObje
...
+        }
+
+        /*******************************************/
+
+        public void BakeGeometry(RhinoDoc doc, ObjectAttributes att, List<Guid> obj_ids)
+        {
+            foreach (Engine.Grasshopper.Objects.GH_BHoMObject item in this.VolatileData.AllData(true))
+            {
+                Guid guid;
+                if (item.BakeGeometry(doc, att, out guid))
```
Comment (adecler) [inline Grasshopper_UI/Components/Parameters/Param_BHoMObject.cs:68]: Probably a good idea to check that `item` is not null

File: Grasshopper_UI/Components/Parameters/Param_IObject.cs [hunk 5]
```diff
@@ -58,6 +62,33 @@ public override bool IsPreviewCapable
...
+        /***************************************************/
+        /**** IGH_BakeAwareObject methods               ****/
+        /***************************************************/
+
+        public void BakeGeometry(RhinoDoc doc, List<Guid> obj_ids)
+        {
+            foreach (Engine.Grasshopper.Objects.GH_IObject item in this.VolatileData.AllData(true))
+            {
+                Guid guid;
+                if (item.BakeGeometry(doc, null, out guid))
```
Comment (adecler) [inline Grasshopper_UI/Components/Parameters/Param_IObject.cs:76]: Probably a good idea to check that `item` is not null

File: Grasshopper_UI/Components/Parameters/Param_IObject.cs [hunk 6]
```diff
@@ -58,6 +62,33 @@ public override bool IsPreviewCapable
...
+        }
+
+        /*******************************************/
+
+        public void BakeGeometry(RhinoDoc doc, ObjectAttributes att, List<Guid> obj_ids)
+        {
+            foreach (Engine.Grasshopper.Objects.GH_IObject item in this.VolatileData.AllData(true))
+            {
+                Guid guid;
+                if (item.BakeGeometry(doc, att, out guid))
```
Comment (adecler) [inline Grasshopper_UI/Components/Parameters/Param_IObject.cs:88]: Probably a good idea to check that `item` is not null

**Review by adecler — CHANGES_REQUESTED** [review]
Brilliant! Thanks @IsakNaslundBh 😄 

Tested both on output param and param components and it works great for the 3 categories (Geometry, IObject & BHoM Object). 

I have never really used bake myself so probably a good idea for someone else to review as well. Otherwise, looks good to me both on code an functionality side with one minor caveat.

File: Grasshopper_UI/Components/Parameters/Param_BHoMGeometry.cs (hunk 1 above)
Comment (IsakNaslundBh) [reply Grasshopper_UI/Components/Parameters/Param_BHoMGeometry.cs:68]: the 'true' in `this.VolatileData.AllData(true)` is supposed to take care of this

image

but can add an extra null check anyway, just to be extra safe if you think that is worth it? :)

Comment (adecler) [reply Grasshopper_UI/Components/Parameters/Param_BHoMGeometry.cs:68]: Ha! Fair enough 😋 

Happy to approve then.

**Review by adecler — APPROVED** [review]

**Review by pawelbaran — APPROVED** [review]
LGTM, works as expected, the only issue I see is the fact that pulled objects need to be cast to `BHoMObject` before they can be baked - this, however, seems to be out of the scope of this PR therefore happy to approve.

**Pull request merged**